File: libs/power-log-parser/src/lib/parsers/game-reset-parser.ts
import { BlockType } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class GameResetParser implements ActionParser {
	readonly ParserName = 'GameResetParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			!this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.Action &&
			(node.Object as Action).Type === (BlockType.GAME_RESET as number)
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			!this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.Action &&
			(node.Object as Action).Type === (BlockType.GAME_RESET as number)
		);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		this.ParserState.PartialReset();
		return [
			GameEventProvider.Create(
				(node.Object as Action).TimeStamp,
				'GAME_RESET_START',
				GameEventHelper.CreateProvider('GAME_RESET_START', null as any, -1, -1, this.StateFacade),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const actionEndTimestamp = action.Data[action.Data.length - 1].TimeStamp;
		return [
			GameEventProvider.Create(
				actionEndTimestamp,
				'GAME_RESET_END',
				GameEventHelper.CreateProvider('GAME_RESET_END', null as any, -1, -1, this.StateFacade),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/GameResetParser.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using System.Collections.Generic;
using System.Linq;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class GameResetParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public GameResetParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && !StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && ((node.Object as Action).Type == (int)BlockType.GAME_RESET);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && !StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && ((node.Object as Action).Type == (int)BlockType.GAME_RESET);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            ParserState.PartialReset();
            return new List<GameEventProvider> { GameEventProvider.Create(
                (node.Object as Action).TimeStamp,
                "GAME_RESET_START",
                GameEvent.CreateProvider(
                    "GAME_RESET_START",
                    null,
                    -1,
                    -1,
                    StateFacade
                ),
                true,
                node
            )};
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var actionTimestamp = action.TimeStamp;
            var actionEndTimestamp = action.Data[action.Data.Count - 1].TimeStamp;
            actionEndTimestamp = actionEndTimestamp.AddMilliseconds(1);
            return new List<GameEventProvider> { GameEventProvider.Create(
                actionEndTimestamp,
                "GAME_RESET_END",
                GameEvent.CreateProvider(
                    "GAME_RESET_END",
                    null,
                    -1,
                    -1,
                    StateFacade
                ),
                true,
                node
            )};
        }
    }
}
